File: .claude/skills/eval-analysis/fidelity.py
#!/usr/bin/env python3
"""
Profile-fidelity cross-validation for Phase 2 memory brackets.

Each memory agent keeps structured per-hand records of the opponent ("villain"):
  - llm-akg-durable : memory-export.json, nodes of type "hand" (body carries
                      "Villain: <c> <c>. Board: <c> ...")
  - llm-md-wiki     : wiki/hands/hand-N.md (frontmatter hand:N, body
                      "**Villain:** <c> <c>", "**Board:** <c> ...")

This script cross-validates those stated reads against engine ground truth in
hands.jsonl. It is fully deterministic — no LLM judging. It scores, per agent:

  - FABRICATION : a specific villain holding stated for a NON-showdown hand.
                  In showdown-only info mode the agent could not have known the
                  opponent's cards, so any such holding was invented. This is the
                  sharpest fidelity-drift signal.
  - CARD ERROR  : villain holding stated for a real showdown hand that does not
                  match the cards actually shown.
  - BOARD ERROR : a claimed board card that was never dealt in that hand.

Each claim is bucketed by hand number so accuracy-vs-hand-number drift is visible.

Usage:
  python3 fidelity.py <experiment-id> [--csv OUT.csv] [--report OUT.md]

Ground truth and session layout are read from the experiment definition under
research/experiments/<id>/<id>.json (groups[].seat0/seat1/seeds) so the script
is reusable across the mdsingle / wiki / fullhistory brackets.
"""

import argparse
import json
import os
import re
import sys

CARD = r"(?:10|[2-9TJQKAtjqka])[shdcSHDC]"
CARD_RE = re.compile(CARD)
REPO = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))


def norm_card(c):
    c = c.replace("10", "T")
    return c[0].upper() + c[1].lower()


def cards_in(text):
    return [norm_card(m.group(0)) for m in CARD_RE.finditer(text)]


# ----- ground truth ----------------------------------------------------------

def load_ground_truth(session_dir):
    gt = {}
    with open(os.path.join(session_dir, "hands.jsonl")) as f:
        for line in f:
            h = json.loads(line)
            n = h["hand_number"]
            gt[n] = {
                "board": [norm_card(c) for c in h.get("board", [])],
                "hole": {int(k): [norm_card(c) for c in v]
                         for k, v in h.get("hole_cards", {}).items()},
                "showdown": bool(h.get("showdown_reached")),
            }
    return gt


# ----- claim extraction ------------------------------------------------------

def extract_akg_claims(agent_dir):
    """Return list of (hand_number, villain_cards|None, board_cards) per hand node."""
    path = os.path.join(agent_dir, "memory-export.json")
    if not os.path.exists(path):
        return []
    data = json.load(open(path))
    claims = []
    for node in data.get("nodes", []):
        if node.get("type") != "hand":
            continue
        m = re.search(r"(\d+)", node.get("id", ""))
        if not m:
            continue
        hn = int(m.group(1))
        body = node.get("body", "")
        claims.append((hn, *parse_villain_board(body)))
    return claims


def extract_wiki_claims(agent_dir):
    hands_dir = os.path.join(agent_dir, "wiki", "hands")
    if not os.path.isdir(hands_dir):
        return []
    claims = []
    for fn in os.listdir(hands_dir):
        if not fn.endswith(".md"):
            continue
        text = open(os.path.join(hands_dir, fn)).read()
        m = re.search(r"hand:\s*(\d+)", text) or re.search(r"hand-(\d+)", fn)
        if not m:
            continue
        hn = int(m.group(1))
        claims.append((hn, *parse_villain_board(text)))
    return claims


def parse_villain_board(text):
    """Pull the villain holding and board off a structured hand record."""
    villain = None
    vm = re.search(r"[Vv]illain[:* ]+\s*(" + CARD + r")\s*[, ]?\s*(" + CARD + r")", text)
    if vm:
        villain = sorted([norm_card(vm.group(1)), norm_card(vm.group(2))])
    # Board: only a "Board:" field that STARTS a line (after optional markdown
    # bullet/bold), capturing the leading contiguous card run. This excludes inline
    # prose ("paired board", or a comparison citing another hand's board) which would
    # otherwise mis-attribute a different hand's cards to this record.
    board = None
    bm = re.search(r"^[\s*_>#-]*[Bb]oard[:*\s]+(" + CARD + r"(?:[ ,]+" + CARD + r")*)",
                   text, re.MULTILINE)
    if bm:
        bc = cards_in(bm.group(1))
        if bc:
            board = bc[:5]
    return villain, board


# ----- verification ----------------------------------------------------------

def verify(claims, gt, villain_seat):
    rows = []
    for hn, villain_cards, board in claims:
        g = gt.get(hn)
        rec = {"hand": hn, "fabrication": False, "card_error": False,
               "board_error": False, "checked_cards": False, "checked_board": False}
        if g is None:
            rec["board_error"] = bool(board)  # references a hand that never happened
            rec["note"] = "no such hand in ground truth"
            rows.append(rec)
            continue
        if villain_cards:
            rec["checked_cards"] = True
            actual = sorted(g["hole"].get(villain_seat, []))
            if not g["showdown"]:
                rec["fabrication"] = True
                rec["note"] = f"villain {villain_cards} stated on non-showdown hand"
            elif villain_cards != actual:
                rec["card_error"] = True
                rec["note"] = f"claimed {villain_cards} vs actual {actual}"
        if board:
            rec["checked_board"] = True
            dealt = g["board"]
            stray = [c for c in board if c not in dealt]
            if stray:
                rec["board_error"] = True
                rec.setdefault("note", "")
                rec["note"] = (rec.get("note", "") + f" board stray {stray} (dealt {dealt})").strip()
        rows.append(rec)
    return rows


# ----- experiment layout -----------------------------------------------------

def load_layout(exp_id):
    path = os.path.join(REPO, "research", "experiments", exp_id, f"{exp_id}.json")
    d = json.load(open(path))
    sessions = []
    for grp in d["groups"]:
        for seed in grp["seeds"]:
            sid = f"{grp['session_base']}-{seed}"
            sessions.append({
                "id": sid,
                "seat0": grp["seat0"],
                "seat1": grp["seat1"],
            })
    return d, sessions


def agg(rows):
    n = len(rows)
    card_claims = sum(r["checked_cards"] for r in rows)
    fab = sum(r["fabrication"] for r in rows)
    cerr = sum(r["card_error"] for r in rows)
    berr = sum(r["board_error"] for r in rows)
    return {"records": n, "card_claims": card_claims, "fabrications": fab,
            "card_errors": cerr, "board_errors": berr}


def drift_buckets(rows, edges=(50, 100)):
    buckets = {}
    bounds = [(0, edges[0]), (edges[0], edges[1]), (edges[1], 10**9)]
    labels = [f"1-{edges[0]}", f"{edges[0]+1}-{edges[1]}", f"{edges[1]+1}+"]
    for (lo, hi), lab in zip(bounds, labels):
        sub = [r for r in rows if lo < r["hand"] <= hi]
        cc = sum(r["checked_cards"] for r in sub)
        buckets[lab] = {
            "records": len(sub),
            "card_claims": cc,
            "fabrications": sum(r["fabrication"] for r in sub),
            "card_errors": sum(r["card_error"] for r in sub),
        }
    return buckets


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("experiment")
    ap.add_argument("--csv")
    ap.add_argument("--report")
    args = ap.parse_args()

    d, sessions = load_layout(args.experiment)
    sess_root = os.path.join(REPO, "research", "experiments", args.experiment, "sessions")

    extractors = {"llm-akg-durable": extract_akg_claims, "llm-md-wiki": extract_wiki_claims}
    per_agent_rows = {}   # agent name -> list of rows (across sessions)
    per_session = []      # (session, agent, agg)
    csv_rows = []

    for s in sessions:
        sdir = os.path.join(sess_root, s["id"])
        if not os.path.isdir(sdir):
            print(f"skip missing session {s['id']}", file=sys.stderr)
            continue
        gt = load_ground_truth(sdir)
        for seat, name in ((0, s["seat0"]), (1, s["seat1"])):
            ext = extractors.get(name)
            if not ext:
                continue
            adir = os.path.join(sdir, "agents", name)
            claims = ext(adir)
            villain_seat = 1 - seat
            rows = verify(claims, gt, villain_seat)
            per_agent_rows.setdefault(name, []).extend(rows)
            per_session.append((s["id"], name, agg(rows)))
            for r in rows:
                csv_rows.append({"session": s["id"], "agent": name, **r})

    # ---- print ----
    print(f"Fidelity cross-validation: {args.experiment}")
    print(f"Strategies: {' vs '.join(sorted(per_agent_rows))}")
    print("\nStructured per-hand records cross-validated against hands.jsonl ground truth.")
    print("(showdown-only info mode: any specific villain holding on a non-showdown hand is a fabrication)\n")

    hdr = f"{'Agent':16} {'recs':>5} {'cardClaims':>10} {'fabricated':>10} {'cardErr':>8} {'boardErr':>8} {'fab%ofCards':>11}"
    print(hdr); print("-" * len(hdr))
    for name in sorted(per_agent_rows):
        a = agg(per_agent_rows[name])
        fabpct = (100.0 * a["fabrications"] / a["card_claims"]) if a["card_claims"] else 0.0
        print(f"{name:16} {a['records']:>5} {a['card_claims']:>10} {a['fabrications']:>10} "
              f"{a['card_errors']:>8} {a['board_errors']:>8} {fabpct:>10.1f}%")

    print("\nFidelity drift (by hand-number bucket):")
    for name in sorted(per_agent_rows):
        b = drift_buckets(per_agent_rows[name])
        print(f"  {name}")
        for lab, v in b.items():
            fabpct = (100.0 * v["fabrications"] / v["card_claims"]) if v["card_claims"] else 0.0
            print(f"    hands {lab:>7}: {v['records']:>3} recs, {v['card_claims']:>3} card-claims, "
                  f"{v['fabrications']:>3} fabricated ({fabpct:.0f}%), {v['card_errors']:>2} card-err")

    print("\nPer session:")
    ph = f"  {'session':16} {'agent':16} {'recs':>4} {'cardClaims':>10} {'fab':>4} {'cardErr':>7} {'boardErr':>8}"
    print(ph)
    for sid, name, a in per_session:
        print(f"  {sid:16} {name:16} {a['records']:>4} {a['card_claims']:>10} "
              f"{a['fabrications']:>4} {a['card_errors']:>7} {a['board_errors']:>8}")

    # ---- csv ----
    if args.csv:
        import csv
        keys = ["session", "agent", "hand", "fabrication", "card_error",
                "board_error", "checked_cards", "checked_board", "note"]
        with open(args.csv, "w", newline="") as f:
            w = csv.DictWriter(f, fieldnames=keys, extrasaction="ignore")
            w.writeheader()
            for r in csv_rows:
                w.writerow(r)
        print(f"\nPer-claim CSV: {args.csv}")

    # ---- markdown report ----
    if args.report:
        with open(args.report, "w") as f:
            f.write(f"# Fidelity Cross-Validation: {args.experiment}\n\n")
            f.write("Objective pass — structured per-hand opponent records cross-validated "
                    "against `hands.jsonl` ground truth. Deterministic, no LLM judging.\n\n")
            f.write("**Fabrication** = a specific villain holding stated for a non-showdown hand "
                    "(unknowable in showdown-only info mode). **Card error** = holding stated for a "
                    "real showdown that doesn't match the cards shown. **Board error** = a claimed "
                    "board card never dealt that hand.\n\n")
            f.write("## Per agent\n\n")
            f.write("| Agent | records | card-claims | fabricated | card-err | board-err | fab % of card-claims |\n")
            f.write("|---|--:|--:|--:|--:|--:|--:|\n")
            for name in sorted(per_agent_rows):
                a = agg(per_agent_rows[name])
                fabpct = (100.0 * a["fabrications"] / a["card_claims"]) if a["card_claims"] else 0.0
                f.write(f"| {name} | {a['records']} | {a['card_claims']} | {a['fabrications']} | "
                        f"{a['card_errors']} | {a['board_errors']} | {fabpct:.1f}% |\n")
            f.write("\n## Fidelity drift (hand-number buckets)\n\n")
            for name in sorted(per_agent_rows):
                f.write(f"### {name}\n\n")
                f.write("| hands | records | card-claims | fabricated | fab % | card-err |\n|---|--:|--:|--:|--:|--:|\n")
                for lab, v in drift_buckets(per_agent_rows[name]).items():
                    fabpct = (100.0 * v["fabrications"] / v["card_claims"]) if v["card_claims"] else 0.0
                    f.write(f"| {lab} | {v['records']} | {v['card_claims']} | {v['fabrications']} | "
                            f"{fabpct:.0f}% | {v['card_errors']} |\n")
                f.write("\n")
            f.write("## Per session\n\n")
            f.write("| session | agent | records | card-claims | fabricated | card-err | board-err |\n|---|---|--:|--:|--:|--:|--:|\n")
            for sid, name, a in per_session:
                f.write(f"| {sid} | {name} | {a['records']} | {a['card_claims']} | "
                        f"{a['fabrications']} | {a['card_errors']} | {a['board_errors']} |\n")
        print(f"Report: {args.report}")


if __name__ == "__main__":
    main()

File: engine/internal/eval/fidelity.go
package eval

import (
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"

	"github.com/RobertGumeny/agent-poker/internal/sessionlog"
)

// Profile-fidelity cross-validation (port of fidelity.py). Fully deterministic,
// no LLM judging: each memory agent's structured per-hand opponent records are
// cross-validated against hands.jsonl ground truth. The extractor is shared
// across substrates (regex over prose) on purpose — giving one substrate
// structured fields would bias the comparison.

const cardPattern = `(?:10|[2-9TJQKAtjqka])[shdcSHDC]`

var (
	cardRE     = regexp.MustCompile(cardPattern)
	villainRE  = regexp.MustCompile(`[Vv]illain[:* ]+\s*(` + cardPattern + `)\s*[, ]?\s*(` + cardPattern + `)`)
	boardRE    = regexp.MustCompile(`(?m)^[\s*_>#-]*[Bb]oard[:*\s]+(` + cardPattern + `(?:[ ,]+` + cardPattern + `)*)`)
	nodeIDRE   = regexp.MustCompile(`(\d+)`)
	wikiHandRE = regexp.MustCompile(`hand:\s*(\d+)`)
	wikiFileRE = regexp.MustCompile(`hand-(\d+)`)
)

// FidelityRow is one cross-validated per-hand claim. Stored in eval.json so the
// experiment report can aggregate across sessions without re-reading raw memory.
type FidelityRow struct {
	Hand         int  `json:"hand"`
	Fabrication  bool `json:"fabrication"`
	CardError    bool `json:"card_error"`
	BoardError   bool `json:"board_error"`
	CheckedCards bool `json:"checked_cards"`
	CheckedBoard bool `json:"checked_board"`
}

type handTruth struct {
	board    []string
	hole     map[int][]string
	showdown bool
}

type fidelityClaim struct {
	hand    int
	villain []string // nil when no villain holding stated
	board   []string // nil when no board stated
}

func normCard(c string) string {
	c = strings.ReplaceAll(c, "10", "T")
	if len(c) < 2 {
		return c
	}
	return strings.ToUpper(c[:1]) + strings.ToLower(c[1:])
}

func cardsIn(text string) []string {
	matches := cardRE.FindAllString(text, -1)
	out := make([]string, 0, len(matches))
	for _, m := range matches {
		out = append(out, normCard(m))
	}
	return out
}

func groundTruthFromHands(hands []sessionlog.HandRecord) map[int]handTruth {
	gt := map[int]handTruth{}
	for _, h := range hands {
		board := make([]string, 0, len(h.Board))
		for _, c := range h.Board {
			board = append(board, normCard(c))
		}
		hole := map[int][]string{}
		for seat, cards := range h.HoleCards {
			norm := make([]string, 0, len(cards))
			for _, c := range cards {
				norm = append(norm, normCard(c))
			}
			hole[seat] = norm
		}
		gt[h.HandNumber] = handTruth{board: board, hole: hole, showdown: h.ShowdownReached}
	}
	return gt
}

func extractAKGClaims(export *MemoryExport) []fidelityClaim {
	if export == nil {
		return nil
	}
	var claims []fidelityClaim
	for _, node := range export.Nodes {
		if node.Type != "hand" {
			continue
		}
		m := nodeIDRE.FindStringSubmatch(node.ID)
		if m == nil {
			continue
		}
		hn, err := strconv.Atoi(m[1])
		if err != nil {
			continue
		}
		villain, board := parseVillainBoard(node.Body)
		claims = append(claims, fidelityClaim{hand: hn, villain: villain, board: board})
	}
	return claims
}

func extractWikiClaims(agentDir string) []fidelityClaim {
	handsDir := filepath.Join(agentDir, "wiki", "hands")
	entries, err := os.ReadDir(handsDir)
	if err != nil {
		return nil
	}
	var claims []fidelityClaim
	for _, entry := range entries {
		fn := entry.Name()
		if entry.IsDir() || !strings.HasSuffix(fn, ".md") {
			continue
		}
		data, err := os.ReadFile(filepath.Join(handsDir, fn))
		if err != nil {
			continue
		}
		text := string(data)
		var hn int
		if m := wikiHandRE.FindStringSubmatch(text); m != nil {
			hn, _ = strconv.Atoi(m[1])
		} else if m := wikiFileRE.FindStringSubmatch(fn); m != nil {
			hn, _ = strconv.Atoi(m[1])
		} else {
			continue
		}
		villain, board := parseVillainBoard(text)
		claims = append(claims, fidelityClaim{hand: hn, villain: villain, board: board})
	}
	return claims
}

// parseVillainBoard pulls the villain holding and board off a structured record.
// The board match must START a line (after optional markdown bullet/bold) so
// inline prose ("paired board", or a citation of another hand) cannot mis-attribute
// cards to this record.
func parseVillainBoard(text string) (villain, board []string) {
	if m := villainRE.FindStringSubmatch(text); m != nil {
		villain = []string{normCard(m[1]), normCard(m[2])}
		sort.Strings(villain)
	}
	if m := boardRE.FindStringSubmatch(text); m != nil {
		bc := cardsIn(m[1])
		if len(bc) > 0 {
			if len(bc) > 5 {
				bc = bc[:5]
			}
			board = bc
		}
	}
	return villain, board
}

func verifyClaims(claims []fidelityClaim, gt map[int]handTruth, villainSeat int) []FidelityRow {
	rows := make([]FidelityRow, 0, len(claims))
	for _, c := range claims {
		rec := FidelityRow{Hand: c.hand}
		g, ok := gt[c.hand]
		if !ok {
			rec.BoardError = len(c.board) > 0 // references a hand that never happened
			rows = append(rows, rec)
			continue
		}
		if len(c.villain) > 0 {
			rec.CheckedCards = true
			actual := append([]string(nil), g.hole[villainSeat]...)
			sort.Strings(actual)
			if !g.showdown {
				rec.Fabrication = true
			} else if !equalStrings(c.villain, actual) {
				rec.CardError = true
			}
		}
		if len(c.board) > 0 {
			rec.CheckedBoard = true
			if hasStray(c.board, g.board) {
				rec.BoardError = true
			}
		}
		rows = append(rows, rec)
	}
	return rows
}

// FidelityAgg is the per-agent aggregate rendered in the report.
type FidelityAgg struct {
	Records      int
	CardClaims   int
	Fabrications int
	CardErrors   int
	BoardErrors  int
}

func aggFidelity(rows []FidelityRow) FidelityAgg {
	a := FidelityAgg{Records: len(rows)}
	for _, r := range rows {
		if r.CheckedCards {
			a.CardClaims++
		}
		if r.Fabrication {
			a.Fabrications++
		}
		if r.CardError {
			a.CardErrors++
		}
		if r.BoardError {
			a.BoardErrors++
		}
	}
	return a
}

// FidelityBucket is a hand-number-range slice of the per-agent aggregate.
type FidelityBucket struct {
	Label string
	FidelityAgg
}

func driftBuckets(rows []FidelityRow) []FidelityBucket {
	type bound struct {
		lo, hi int
		label  string
	}
	bounds := []bound{{0, 50, "1-50"}, {50, 100, "51-100"}, {100, 1 << 30, "101+"}}
	buckets := make([]FidelityBucket, 0, len(bounds))
	for _, b := range bounds {
		var sub []FidelityRow
		for _, r := range rows {
			if r.Hand > b.lo && r.Hand <= b.hi {
				sub = append(sub, r)
			}
		}
		buckets = append(buckets, FidelityBucket{Label: b.label, FidelityAgg: aggFidelity(sub)})
	}
	return buckets
}

// fidelityExtractorFor returns the claim extractor for a given agent strategy, or
// nil for substrates with no machine-extractable structured records.
func fidelityRows(agent AgentArtifacts, gt map[int]handTruth) []FidelityRow {
	var claims []fidelityClaim
	switch agent.Seat.Name {
	case "llm-akg-durable":
		claims = extractAKGClaims(agent.MemoryExport)
	case "llm-md-wiki":
		claims = extractWikiClaims(agent.Dir)
	default:
		return nil
	}
	if claims == nil {
		return nil
	}
	return verifyClaims(claims, gt, 1-agent.Seat.Seat)
}

func equalStrings(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func hasStray(claimed, dealt []string) bool {
	for _, c := range claimed {
		found := false
		for _, d := range dealt {
			if c == d {
				found = true
				break
			}
		}
		if !found {
			return true
		}
	}
	return false
}
